src/app/clusters/basic-information/CodegenIntegration.cpp

```diff
@@ -0,0 +1,96 @@
+#include <app/util/attribute-storage.h>
+#include <app/util/endpoint-config-api.h>
+#include <data-model-providers/codegen/CodegenDataModelProvider.h>
+#include <lib/support/BitFlags.h>
+
+using namespace chip;
+using namespace chip::app;
+using namespace chip::app::Clusters;
+using namespace chip::app::Clusters::BasicInformation::Attributes;
+using namespace chip::app::Clusters::BasicInformation::StaticApplicationConfig;
+using chip::Protocols::InteractionModel::Status;
+
+namespace {
+
+// BasicInformationCluster implementation is specifically implemented
+// only for the root endpoint (endpoint 0)
+// So either:
+//   - we have a fixed config and it is endpoint 0 OR
+//   - we have a fully dynamic config
+
+static constexpr size_t kBasicInformationFixedClusterCount = BasicInformation::StaticApplicationConfig::kFixedClusterConfig.size();
+
+static_assert((kBasicInformationFixedClusterCount == 0) ||
+                  ((kBasicInformationFixedClusterCount == 1) &&
+                   BasicInformation::StaticApplicationConfig::kFixedClusterConfig[0].endpointNumber == kRootEndpointId),
+              "Basic Information cluster MUST be on endpoint 0");
+
+ServerClusterRegistration gRegistration(BasicInformationCluster::Instance());
+
+} // namespace
+
+void emberAfBasicInformationClusterInitCallback(EndpointId endpointId)
+{
+    VerifyOrReturn(endpointId == kRootEndpointId);
+
+    BasicInformationCluster::Instance().OptionalAttributes() =
+        BitFlags<OptionalBasicInformationAttributes>()
+            .Set(OptionalBasicInformationAttributes::kManufacturingDate,
+                 emberAfContainsAttribute(endpointId, BasicInformation::Id, ManufacturingDate::Id))
+            .Set(OptionalBasicInformationAttributes::kPartNumber,
+                 emberAfContainsAttribute(endpointId, BasicInformation::Id, PartNumber::Id))
+            .Set(OptionalBasicInformationAttributes::kProductURL,
+                 emberAfContainsAttribute(endpointId, BasicInformation::Id, ProductURL::Id))
+            .Set(OptionalBasicInformationAttributes::kProductLabel,
+                 emberAfContainsAttribute(endpointId, BasicInformation::Id, ProductLabel::Id))
+            .Set(OptionalBasicInformationAttributes::kSerialNumber,
+                 emberAfContainsAttribute(endpointId, BasicInformation::Id, SerialNumber::Id))
+            .Set(OptionalBasicInformationAttributes::kLocalConfigDisabled,
+                 emberAfContainsAttribute(endpointId, BasicInformation::Id, LocalConfigDisabled::Id))
+            .Set(OptionalBasicInformationAttributes::kReachable,
+                 emberAfContainsAttribute(endpointId, BasicInformation::Id, Reachable::Id))
+            .Set(OptionalBasicInformationAttributes::kProductAppearance,
+                 emberAfContainsAttribute(endpointId, BasicInformation::Id, ProductAppearance::Id))
+            // This is NOT typical, however we try to respect ZAP here. MCORE_FS tests require this
```

Comment:
Why do they require this?  This piece feels like it needs a much more extensive explanation.

Reply:
added explanation in several places. Also fixed revision (at the cost of minor flash probably): we should report an old revision when we do not provide the unique ID.